### PR TITLE
[Stdlib] Use direct memcmp in startswith() and endswith()

### PR DESCRIPTION
**Pull request by msaelices**
Both methods previously used `find()`/`rfind()` which scanned the entire string. Now they compare only the prefix/suffix bytes via `memcmp`, reducing from O(n) to O(m) where m is the prefix/suffix length.

### CODE REVIEW

File: mojo/stdlib/std/collections/string/string_slice.mojo [hunk 1]
```diff
@@ -1809,14 +1812,18 @@ struct StringSlice[mut: Bool, //, origin: Origin[mut=mut]](
...
-        # FIXME: use normalize_index
-        return StringSlice[Self.origin](
-            ptr=self.unsafe_ptr() + start, length=end - start
-        ).endswith(suffix)
+        return (
+            memcmp(
+                self.unsafe_ptr() + e - len(suffix),
+                suffix.unsafe_ptr(),
+                len(suffix),
+            )
```
[the remark this answers is not in the document]
Comment (JoeLoser) [reply mojo/stdlib/std/collections/string/string_slice.mojo]: +1, we need some extra tests here. Some examples:

```mojo
# Out-of-range indices (should clamp, not crash)
assert_true("hello".startswith("hello", 0, 100))  # end beyond length
assert_false("hello".startswith("x", 10))         # start beyond length

# start > end
assert_false("hello".startswith("h", 3, 1))       # invalid range
```

[the remark this answers is not in the document]
Comment (msaelices) [reply mojo/stdlib/std/collections/string/string_slice.mojo]: Thanks for the suggestions. These tests actually triggered an issue in the logic, fixed here (with tests): https://github.com/msaelices/modular/commit/f809edfb300a72517a51fdb91226a6c2fd6a1b7b

File: mojo/stdlib/std/collections/string/string_span.mojo [hunk 2]
```diff
@@ -1784,12 +1784,15 @@ struct StringSlice[mut: Bool, //, origin: Origin[mut=mut]](
...
-        if end == -1:
-            return self.find(prefix, start) == start
-        # FIXME: use normalize_index
-        return StringSlice[Self.origin](
-            ptr=self.unsafe_ptr() + start, length=end - start
-        ).startswith(prefix)
+        var n = len(self)
+        var s = start if start >= 0 else max(0, start + n)
+        var e = n if end == -1 else min(n, end if end >= 0 else max(0, end + n))
+        if s > e or len(prefix) > e - s:
```
Comment (JoeLoser) [inline mojo/stdlib/std/collections/string/string_span.mojo]: The `end == -1` sentinel conflicts with `-1` as a legitimate negative index. In Python:

```python
>>> "hello".endswith("o", 0, -1)   # slice [0:4]="hell" → False
>>> "hello".endswith("l", 0, -1)   # slice [0:4]="hell" → True
```

But this code treats `end == -1` as `e = n` (full string), so both would return the wrong result. This is a pre-existing bug (the old code had it too), so not a regression — but since you're rewriting the index normalization anyway, this is the ideal time to fix it.

Possible fix: use a different sentinel (e.g., `Int.MAX` or `Optional[Int]`), or handle `-1` the same as other negatives and use a separate signal for "default."

File: mojo/stdlib/test/collections/string/test_string_span.mojo [hunk 3]
```diff
@@ -862,6 +862,14 @@ def test_startswith():
     assert_true(ab.startswith("a", end=1))
     assert_true(ab.startswith("ab"))
 
+    # Out-of-range indices should clamp, not crash.
+    var hello = StringSlice("hello")
+    assert_true(hello.startswith("hello", 0, 100))
+    assert_false(hello.startswith("x", 10))
+    assert_false(hello.startswith("h", 3, 1))
```
Comment (JoeLoser) [inline mojo/stdlib/test/collections/string/test_string_span.mojo:918]: Nice edge-case coverage for negative `start` and out-of-range `end`. Consider also adding negative `end` tests — these would exercise the `end == -1` sentinel path and other negative-end normalization:

```mojo
# Negative end values
assert_true(hello.startswith("hel", 0, -2))    # slice [0:3]
assert_true(hello.startswith("ell", -4, -1))   # slice [1:4]
assert_false(hello.startswith("hello", 0, -1)) # slice [0:4], too short
```

(And equivalents for `endswith` below.)

**Review by JoeLoser — APPROVED** [review]
The core optimization is correct and very much welcomed — scanning the full string to check a prefix/suffix was wasteful. The index normalization is strictly better than the old code. A couple of suggestions below.

**Comment by msaelices** [comment]
@JoeLoser this PR was approved long ago, and I think it would be still a good optimization. WDYT?
